n3fit/src/n3fit/backends/keras_backend/MetaModel.py

```diff
@@ -53,121 +53,87 @@ def _default_loss(y_true, y_pred): # pylint: disable=unused-argument
-        # *this option was dropped at some point by TF, the code below keeps this behaviour
-        # We will store within the model the following quantities:
-        #   -> x_in: arrays containing the x-input to the model
-        #   -> tensors_in: when the x-input is known at compile time, we store a reference to the tensor
-        # We pass TensorFlow a dictionary {k: tensor} containing placeholders which will be automatically filled
-        # whenever x_in/tensor_in is known at compile time
+        #input_list = input_tensors
+        #output_list = output_tensors
+
+        if input_values is None:
+            input_values = {}
 
         x_in = {}
-        tensors_in = {}
-        input_dict = {}
-        for input_tensor in input_list:
-            # If the input contains a tensor_content, store it to use at predict/fit/eval times
-            # otherwise, put a placeholder None as it will come from the outside
-            name = input_tensor.name.rsplit(":", 1)[0]
-            input_dict[name] = input_tensor
-            try:
-                x_in[name] = op.numpy_to_tensor(input_tensor.tensor_content)
-                tensors_in[name] = input_tensor
-            except AttributeError:
-                x_in[name] = None
-                tensors_in[name] = None
-
-        super().__init__(input_dict, output_list, **kwargs)
+        # Go over the inputs. If we can deduce a constant value, either because
+        # it is set in input_values or because it has a tensor_content, we
+        # store it. Otherwise we mark the input as required when making
+        # predictions.
+        for k, v in input_tensors.items():
+            if k in input_values:
+                x_in[k] = input_values[k]
+            elif hasattr(v, "tensor_content"):
+                x_in[k] = v.tensor_content
+            else:
+                self.required_slots.add(k)
+        self.tensors_in = input_tensors
+        super().__init__(input_tensors, output_tensors, **kwargs)
 
         self.x_in = x_in
-        self.tensors_in = tensors_in
 
         self.target_tensors = None
         self.compute_losses_function = None
         self._scaler = scaler
 
+    @tf.autograph.experimental.do_not_convert
     def _parse_input(self, extra_input=None):
         """Returns the input data the model was compiled with.
         Introduces the extra_input in the places asigned to the placeholders.
 
         If the model was generated with a scaler, the input will be scaled accordingly
         """
-        if isinstance(extra_input, dict) and self._scaler is not None:
+        if extra_input is None:
+            if self.required_slots:
+                raise ValueError(
+                    f"The following inputs must be provided: {self.required_slots}"
+                )
+            return self.x_in
+
+        if diff := (self.required_slots - extra_input.keys()):
+            raise ValueError(f"The following inputs must be provided {diff}")
```

Comment:
A few comments here:

- `extra_input` could enter as a list (or as whatever without `keys`). Since we are checking for mistakes, this should be tested for (and the error message should also include the keys that are required).
- The `diff` should also check whether there are `extra_inputs` being provided that is not required.